File: php/3.1/Pubnub.php
<?php

/**
 * PubNub 3.0 Real-time Push Cloud API
 *
 * @author Stephen Blum
 * @package Pubnub
 */
class Pubnub {
    private $ORIGIN        = 'pubsub.pubnub.com';
    private $PUBLISH_KEY   = 'demo';
    private $SUBSCRIBE_KEY = 'demo';
    private $SECRET_KEY    = false;
    private $SSL           = false;

    /**
     * Pubnub
     *
     * Init the Pubnub Client API
     *
     * @param string $publish_key required key to send messages.
     * @param string $subscribe_key required key to receive messages.
     * @param string $secret_key optional key to sign messages.
     * @param string $origin optional setting for cloud origin.
     * @param boolean $ssl required for 2048 bit encrypted messages.
     */
    function Pubnub(
        $publish_key   = 'demo',
        $subscribe_key = 'demo',
        $secret_key    = false,
        $ssl           = false,
        $origin        = false
    ) {
        $this->PUBLISH_KEY   = $publish_key;
        $this->SUBSCRIBE_KEY = $subscribe_key;
        $this->SECRET_KEY    = $secret_key;
        $this->SSL           = $ssl;

        if ($origin) $this->ORIGIN = $origin;

        if ($ssl) $this->ORIGIN = 'https://' . $this->ORIGIN;
        else      $this->ORIGIN = 'http://'  . $this->ORIGIN;
    }

    /**
     * Publish
     *
     * Send a message to a channel.
     *
     * @param array $args with channel and message.
     * @return array success information.
     */
    function publish($args) {
        ## Fail if bad input.
        if (!($args['channel'] && $args['message'])) {
            echo('Missing Channel or Message');
            return false;
        }

        ## Capture User Input
        $channel = $args['channel'];
        $message = json_encode($args['message']);

        ## Generate String to Sign
        $string_to_sign = implode( '/', array(
            $this->PUBLISH_KEY,
            $this->SUBSCRIBE_KEY,
            $this->SECRET_KEY,
            $channel,
            $message
        ) );

        ## Sign Message
        $signature = $this->SECRET_KEY ? md5($string_to_sign) : '0';

        ## Send Message
        return $this->_request(array(
            'publish',
            $this->PUBLISH_KEY,
            $this->SUBSCRIBE_KEY,
            $signature,
            $channel,
            '0',
            $message
        ));
    }

    /**
     * Subscribe
     *
     * This is BLOCKING.
     * Listen for a message on a channel.
     *
     * @param array $args with channel and message.
     * @return mixed false on fail, array on success.
     */
    function subscribe($args) {
        ## Capture User Input
        $channel   = $args['channel'];
        $callback  = $args['callback'];
        $timetoken = isset($args['timetoken']) ? $args['timetoken'] : '0';

        ## Fail if missing channel
        if (!$channel) {
            echo("Missing Channel.\n");
            return false;
        }

        ## Fail if missing callback
        if (!$callback) {
            echo("Missing Callback.\n");
            return false;
        }

        ## Begin Recusive Subscribe
        try {
            ## Wait for Message
            $response = $this->_request(array(
                'subscribe',
                $this->SUBSCRIBE_KEY,
                $channel,
                '0',
                $timetoken
            ));

            $messages          = $response[0];
            $args['timetoken'] = $response[1];

            ## If it was a timeout
            if (!count($messages)) {
                return $this->subscribe($args);
            }

            ## Run user Callback and Reconnect if user permits.
            foreach ($messages as $message) {
                if (!$callback($message)) return;
            }

            ## Keep Listening.
            return $this->subscribe($args);
        }
        catch (Exception $error) {
            sleep(1);
            return $this->subscribe($args);
        }
    }

    /**
     * History
     *
     * Load history from a channel.
     *
     * @param array $args with 'channel' and 'limit'.
     * @return mixed false on fail, array on success.
     */
    function history($args) {
        ## Capture User Input
        $limit   = +$args['limit'] ? +$args['limit'] : 10;
        $channel = $args['channel'];

        ## Fail if bad input.
        if (!$channel) {
            echo('Missing Channel');
            return false;
        }

        ## Get History
        return $this->_request(array(
            'history',
            $this->SUBSCRIBE_KEY,
            $channel,
            '0',
            $limit
        ));
    }

    /**
     * Time
     *
     * Timestamp from PubNub Cloud.
     *
     * @return int timestamp.
     */
    function time() {
        ## Get History
        $response = $this->_request(array(
            'time',
            '0'
        ));

        return $response[0];
    }

    /**
     * Request URL
     *
     * @param array $request of url directories.
     * @return array from JSON response.
     */
    private function _request($request) {
        $request = array_map( 'Pubnub::_encode', $request );
        array_unshift( $request, $this->ORIGIN );

        $ctx = stream_context_create(array(
            'http' => array( 'timeout' => 200 ) 
        ));

        return json_decode( @file_get_contents(
            implode( '/', $request ), 0, $ctx
        ), true );
    }

    /**
     * Encode
     *
     * @param string $part of url directories.
     * @return string encoded string.
     */
    private static function _encode($part) {
        return implode( '', array_map(
            'Pubnub::_encode_char', str_split($part)
        ) );
    }

    /**
     * Encode Char
     *
     * @param string $char val.
     * @return string encoded char.
     */
    private static function _encode_char($char) {
        if (strpos( ' ~`!@#$%^&*()+=[]\\{}|;\':",./<>?', $char ) === false)
            return $char;
        return rawurlencode($char);
    }
}


?>

File: ruby/3.1/lib/pubnub.rb
## www.pubnub.com - PubNub realtime push service in the cloud.
## http://www.pubnub.com/blog/ruby-push-api - Ruby Push API Blog

## PubNub Real Time Push APIs and Notifications Framework
## Copyright (c) 2010 Stephen Blum
## http://www.pubnub.com/

## -----------------------------------
## PubNub 3.1 Real-time Push Cloud API
## -----------------------------------

## including required libraries
require 'openssl'
require 'base64'
require 'open-uri'
require 'uri'
require 'net/http'
require 'net/https'
require 'json'
require 'pp'
require 'rubygems'
require 'securerandom'
require 'digest'
require 'pubnub_crypto'

class Pubnub
  MAX_RETRIES = 3
  retries=0
  #**
  #* Pubnub 3.1 with Cipher Key
  #*
  #* Init the Pubnub Client API
  #*
  #* @param string publish_key required key to send messages.
  #* @param string subscribe_key required key to receive messages.
  #* @param string secret_key required key to sign messages.
  #* @param string cipher_key required to encrypt messages.
  #* @param boolean ssl required for 2048 bit encrypted messages.
  #*
  def initialize( publish_key, subscribe_key, secret_key, cipher_key, ssl_on )
    @publish_key   = publish_key
    @subscribe_key = subscribe_key
    @secret_key    = secret_key
    @cipher_key    = cipher_key
    @ssl           = ssl_on
    @origin        = 'pubsub.pubnub.com'

    if @ssl
      @origin = 'https://' + @origin
    else
      @origin = 'http://'  + @origin
    end
  end

  #**
  #* Publish
  #*
  #* Send a message to a channel.
  #*
  #* @param array args with channel and message.
  #* @return array success information.
  #*
  def publish(args)
    ## Fail if bad input.
    if !(args['channel'] && args['message'] && args['callback'])
      puts('Missing Channel or Message or Callback')
      return false
    end

    ## Capture User Input
    channel = args['channel']
    message = args['message']
    callback = args['callback']

    # Encryption of message
    if @cipher_key.length > 0
      pc=PubnubCrypto.new(@cipher_key)
      if message.is_a? Array
        message=pc.encryptArray(message)
      else
        message=pc.encryptObject(message)
      end
    else
      message = args['message'].to_json();
    end

    ## Sign message using HMAC
    String signature = '0'
    if @secret_key.length > 0
      signature = "{@publish_key,@subscribe_key,@secret_key,channel,message}"
      digest = OpenSSL::Digest.new("sha256")
      key = [ @secret_key ]
      hmac = OpenSSL::HMAC.hexdigest(digest, key.pack("H*"), signature)
      signature = hmac
    end

    ## Send Message
    request = [ 'publish', @publish_key, @subscribe_key, signature, channel, '0', message ]
    args['request'] = request
    args['callback'] = callback
    _request(args)
  end

  #**
  #* Subscribe
  #*
  #* This is NON-BLOCKING.
  #* Listen for a message on a channel.
  #*
  #* @param array args with channel and message.
  #* @return false on fail, array on success.
  #*
  def subscribe(args)
    ## Capture User Input
    channel   = args['channel']
    callback  = args['callback']

    ## Fail if missing channel
    if !channel
      puts "Missing Channel."
      return false
    end

    ## Fail if missing callback
    if !callback
      puts "Missing Callback."
      return false
    end

    ## EventMachine loop
    #EventMachine.run do
      timetoken = 0
      request = [ 'subscribe', @subscribe_key, channel, '0', timetoken.to_s ]
      args['request'] = request
      _subscribe(args)
    #end
  end

  #**
  #* History
  #*
  #* Load history from a channel.
  #*
  #* @param array args with 'channel' and 'limit'.
  #* @return mixed false on fail, array on success.
  #*
  def history(args)
    ## Capture User Input
    limit   = +args['limit'] ? +args['limit'] : 5
    channel = args['channel']
    callback = args['callback']

    ## Fail if bad input.
    if (!channel)
      puts 'Missing Channel.'
      return false
    end
    if (!callback)
      puts 'Missing Callback.'
      return false
    end

    ## Get History
    request = [ 'history', @subscribe_key, channel, '0', limit.to_s ]
    args['request'] = request
    _request(args)
  end

  #**
  #* Time
  #*
  #* Timestamp from PubNub Cloud.
  #*
  #* @return int timestamp.
  #*
  def time(args)
    request = [ 'time', '0' ]
    args['request'] = request
    _request(args)
  end

  #**
  #* UUID
  #*
  #* Unique identifier generation
  #*
  #* @return Unique Identifier
  #*
  def UUID()
    uuid=SecureRandom.base64(32).gsub("/","_").gsub(/=+$/,"")
  end

  private

  #**
  #* Request URL for subscribe
  #*
  #* @param array request of url directories.
  #* @return array from JSON response.
  #*
  def _subscribe(args)
    channel = args['channel']
    callback = args['callback']
    request = args['request']

    # Construct Request
    url = encode_URL(request);
    url = @origin + url

    # Execute Request
    loop do
      begin

        open(url, 'r', :read_timeout => 300) do |f|
          http_response = JSON.parse(f.read)
          messages = http_response[0]
          timetoken = http_response[1]

          next if !messages.length

          ## Run user Callback and Reconnect if user permits.
          ## Capture the message and encrypt it
          if @cipher_key.length > 0
            pc = PubnubCrypto.new(@cipher_key)
            messages.each do |message|
              if message.is_a? Array
                message=pc.decryptArray(message)
              else
                message=pc.decryptObject(message)
              end
              if !callback.call(message)
                return
              end
            end
          else
            messages.each do |message|
              if !callback.call(message)
                return
              end
            end
          end

          request = ['subscribe', @subscribe_key, channel, '0', timetoken.to_s]
          args['request'] = request
          # Recusive call to _subscribe
          _subscribe(args)
          
        end

      rescue Timeout::Error => e
        logger.debug "Caught #{e.message}, restarting connection."
        retry

      end

    end

  end

  #**
  #* Request URL
  #*
  #* @param array request of url directories.
  #* @return array from JSON response.
  #*
  def _request(args)
    request  = args['request']
    callback = args['callback']
    url = encode_URL(request)
    url = @origin + url
      open(url) do |f|
        response = JSON.parse(f.read)
        if request[0] == 'history'
          if @cipher_key.length > 0
            myarr=Array.new()
            response.each do |message|
              pc=PubnubCrypto.new(@cipher_key)
              if message.is_a? Array
                message=pc.decryptArray(message)
              else
                message=pc.decryptObject(message)
              end
              myarr.push(message)
            end
            callback.call(myarr)
          else
            callback.call(response)
          end
        elsif request[0] == 'publish'
          callback.call(response)
        else
          callback.call(response)
        end
      end
  end

  def encode_URL(request)
    ## Construct Request URL
    url = '/' + request.map{ |bit| bit.split('').map{ |ch|
        ' ~`!@#$%^&*()+=[]\\{}|;\':",./<>?'.index(ch) ?
        '%' + ch.unpack('H2')[0].to_s.upcase : URI.encode(ch)
      }.join('') }.join('/')
    return url
  end
end
